server/src/main/java/org/elasticsearch/action/bulk/TransportShardBulkAction.java

```diff
@@ -114,157 +111,213 @@ protected boolean resolveIndex() {
                 }
 
                 @Override
                 public void onTimeout(TimeValue timeout) {
-                    waitingFuture.onFailure(
-                        new MapperException("timed out while waiting for a dynamic mapping update"));
+                    mappingUpdateListener.onFailure(new MapperException("timed out while waiting for a dynamic mapping update"));
                 }
-            });
-            waitingFuture.get();
-        };
-        return performOnPrimary(request, primary, updateHelper, threadPool::absoluteTimeInMillis,
-            new ConcreteMappingUpdatePerformer(), waitForMappingUpdate);
+            }), listener, threadPool
+        );
     }
 
-    public static WritePrimaryResult<BulkShardRequest, BulkShardResponse> performOnPrimary(
+    public static void performOnPrimary(
         BulkShardRequest request,
         IndexShard primary,
         UpdateHelper updateHelper,
         LongSupplier nowInMillisSupplier,
         MappingUpdatePerformer mappingUpdater,
-        CheckedRunnable<Exception> waitForMappingUpdate) throws Exception {
-        BulkPrimaryExecutionContext context = new BulkPrimaryExecutionContext(request, primary);
-        return performOnPrimary(context, updateHelper, nowInMillisSupplier, mappingUpdater, waitForMappingUpdate);
-    }
-
-    private static WritePrimaryResult<BulkShardRequest, BulkShardResponse> performOnPrimary(
-        BulkPrimaryExecutionContext context, UpdateHelper updateHelper, LongSupplier nowInMillisSupplier,
-        MappingUpdatePerformer mappingUpdater, CheckedRunnable<Exception> waitForMappingUpdate) throws Exception {
-
-        while (context.hasMoreOperationsToExecute()) {
-            executeBulkItemRequest(context, updateHelper, nowInMillisSupplier, mappingUpdater, waitForMappingUpdate);
-            assert context.isInitial(); // either completed and moved to next or reset
-        }
-        return new WritePrimaryResult<>(context.getBulkShardRequest(), context.buildShardResponse(), context.getLocationToSync(),
-            null, context.getPrimary(), logger);
+        Consumer<ActionListener<Void>> waitForMappingUpdate,
+        ActionListener<PrimaryResult<BulkShardRequest, BulkShardResponse>> listener,
+        ThreadPool threadPool) {
+        new ActionRunnable<PrimaryResult<BulkShardRequest, BulkShardResponse>>(listener) {
+
+            private final Executor executor = threadPool.executor(ThreadPool.Names.WRITE);
+
+            private final BulkPrimaryExecutionContext context = new BulkPrimaryExecutionContext(request, primary);
+
+            @Override
+            protected void doRun() {
+                while (context.hasMoreOperationsToExecute()) {
+                    if (executeBulkItemRequest(context, updateHelper, nowInMillisSupplier, mappingUpdater, waitForMappingUpdate,
+                        ActionListener.wrap(v -> executor.execute(this), listener::onFailure)) == false) {
```

Comment:
We are moving from the `update` to the `write` thread here after a mapping update updates the cluster state. I'm a little unsure about this tbh. Is it a problem that when we're under pressure and could get a bulk rejection that the mapping update could have happened but the write then got rejected?
I assumed not, as the same kind of inconsistency can happen if the ES process dies after the mapping update in the current implementation, but thought I should point that out (obviously this is a much more likely spot now when bulk rejections happen).

Reply:
The main reason an earlier refactoring of this class did not move this bit to async code was that this part here is problematic. We might have partially executed the batch and generated sequence numbers for it. Failing the request here now is not an option as it will otherwise have the replica go out of sync with the primary, lead to gaps in the history on the replica and block local and global checkpoint advancement, given that the partially executed request won't be replicated.
There are two options to explore: 1) forcing this onto the write executor or 2) marking the current and all subsequent bulk items as failed, and then replicate the request normally to the replicas. I would be interested in @bleskes's thoughts on this.

Also this will require extensive testing given the brittle nature of it.

Reply:
Indeed we avoided doing this in order to not have to deal with rejections / not to have a way bypass queue sizes (which will happen with force execution) at the time. I think the added value of the extra testing coverage merits exploring our options here. I'm inclined to go with the second suggestion (mark the rest of the operations as failed on rejection). I'm not too worried about having committed the mapping changes but ending up not using them. I think that just an edge case and as Armin mentioned - it can happen anyway.

Reply:
@ywelsch @bleskes thanks for taking a look. I implemented option 2 real quick in https://github.com/elastic/elasticsearch/pull/39793/commits/f3b59c2730edb7cf9035e4339673e22eee258b59 (on bulk reject just loop through all remaining requests and fail them). If this looks ok, I'd look into adding test coverage for the scenario mentioned above?

Reply:
I think I was also able to write a valid test (at least I reproduced the issue that @ywelsch mentioned) in https://github.com/elastic/elasticsearch/pull/39793/commits/a3c8cac75e1340e7eca97f91efa57cef0fde709c. Take a look when you have a sec :)